lapce-app/src/editor/view.rs

```diff
@@ -1268,241 +1424,60 @@ fn get_sticky_header_info(
-                        .buffer
-                        .with_untracked(|b| b.line_of_offset(preedit.offset));
-                    if preedit_line == line {
-                        col + *start
-                    } else {
-                        col
-                    }
-                } else {
-                    col
-                }
-            })
-        })
-        .unwrap_or(col);
 
-    let x0 = view.line_point_of_line_col(line, col, 12).x;
-    if block {
-        let right_offset = view.move_right(offset, Mode::Insert, 1);
-        let (_, right_col) = view.offset_to_line_col(right_offset);
-        let x1 = view.line_point_of_line_col(line, right_col, 12).x;
-
-        let width = if x1 > x0 { x1 - x0 } else { 7.0 };
-        CursorRender::Caret { x: x0, width, line }
-    } else {
-        CursorRender::Caret {
-            x: x0 - 1.0,
-            width: 2.0,
-            line,
-        }
-    }
-}
-
-#[allow(clippy::too_many_arguments)]
-fn visual_cursor(
-    view: &EditorViewData,
-    start: usize,
-    end: usize,
-    mode: &VisualMode,
-    horiz: Option<&ColPosition>,
-    char_width: f64,
-    is_active: bool,
-    screen_lines: &ScreenLines,
-) -> Vec<CursorRender> {
-    let (start_line, start_col) = view.offset_to_line_col(start.min(end));
-    let (end_line, end_col) = view.offset_to_line_col(start.max(end));
-    let (cursor_line, _) = view.offset_to_line_col(end);
-
-    let mut renders = Vec::new();
-
-    for line in &screen_lines.lines {
-        let line = *line;
-        if line < start_line {
-            continue;
-        }
-
-        if line > end_line {
-            break;
-        }
-
-        let left_col = match mode {
-            VisualMode::Normal => {
-                if start_line == line {
-                    start_col
-                } else {
-                    0
-                }
-            }
-            VisualMode::Linewise => 0,
-            VisualMode::Blockwise => {
-                let max_col = view.line_end_col(line, false);
-                let left = start_col.min(end_col);
-                if left > max_col {
-                    continue;
-                }
-                left
-            }
-        };
+    let col = col
+        + doc
+            .preedit
+            .with_untracked(|preedit| {
+                preedit.as_ref().and_then(|preedit| {
+                    preedit.cursor.as_ref().and_then(|&(start, _)| {
+                        let preedit_line = doc
+                            .buffer
+                            .with_untracked(|b| b.line_of_offset(preedit.offset));
 
-        let (right_col, line_end) = match mode {
-            VisualMode::Normal => {
-                if line == end_line {
-                    let max_col = view.line_end_col(line, true);
+                        (preedit_line == line).then_some(start)
+                    })
+                })
+            })
+            .unwrap_or(0);
```

Comment:
I'm not sure what a preedit is or what this logic is for.

Reply:
preedit is for IME

Reply:
I've done some testing entering emoji with IBus and cursor placement seems to work as well as it did prior to this PR.

IME doesn't work well with multiple cursors - only one cursor displays the preedit text, and with multiple cursors on the same line, the cursors other than the one displaying the preedit text render in the wrong place. These issues were both present prior to this PR though, so I think they could be fixed in a separate PR.